proofs.go

```diff
@@ -1,93 +1,168 @@
+	// of the key being proven, in the case of a non-membership proof.
+	NonMembershipLeafData []byte
+
+	// BitMask, in the case of a compact proof, is a bit mask of the sidenodes
+	// of the proof where an on-bit indicates that the sidenode at the bit's
+	// index is a placeholder. This is only set if the proof is compact.
+	BitMask []byte
+
+	// NumSideNodes, in the case of a compact proof, indicates the number of
+	// sidenodes in the proof when decompacted. This is only set if the proof is compact.
+	NumSideNodes int
+}
+
+func (proof *SparseMerkleProof) sanityCheck(th *treeHasher) bool {
+	// Do a basic sanity check on the proof, so that a malicious proof cannot
+	// cause the verifier to fatally exit (e.g. due to an index out-of-range
+	// error) or cause a CPU DoS attack.
+
+	// Check that the number of supplied sidenodes does not exceed the maximum possible.
+	if len(proof.SideNodes) > th.pathSize()*8 ||
+
+		// Check that leaf data for non-membership proofs is the correct size.
+		(proof.NonMembershipLeafData != nil && len(proof.NonMembershipLeafData) != len(leafPrefix)+th.pathSize()+th.hasher.Size()) ||
+
+		// Compact proofs: check that NumSideNodes is within the right range.
+		proof.NumSideNodes < 0 ||
+		proof.NumSideNodes > th.pathSize()*8 ||
+
+		// Compact proofs: check that the length of the bit mask is as expected
+		// according to NumSideNodes.
+		len(proof.BitMask) != int(math.Ceil(float64(proof.NumSideNodes)/float64(8))) ||
+
+		// Compact proofs: check that the correct number of sidenodes have been
+		// supplied according to the bit mask.
+		(proof.NumSideNodes > 0 && len(proof.SideNodes) != proof.NumSideNodes-countSetBits(proof.BitMask)) {
+		return false
+	}
+
+	// Check that all supplied sidenodes are the correct size.
+	for _, v := range proof.SideNodes {
+		if len(v) != th.hasher.Size() {
+			return false
+		}
+	}
+
+	return true
+}
+
 // VerifyProof verifies a Merkle proof.
-func VerifyProof(proof [][]byte, root []byte, key []byte, value []byte, hasher hash.Hash) bool {
-    hasher.Write(key)
-    path := hasher.Sum(nil)
-    hasher.Reset()
-
-    hasher.Write(value)
-    currentHash := hasher.Sum(nil)
-    hasher.Reset()
-
-    if len(proof) != hasher.Size() * 8 {
-        return false
-    }
-
-    for i := hasher.Size() * 8 - 1; i >= 0; i-- {
-        node := make([]byte, hasher.Size())
-        copy(node, proof[i])
-        if len(node) != hasher.Size() {
-            return false
-        }
-        if hasBit(path, i) == right {
-            hasher.Write(append(node, currentHash...))
-            currentHash = hasher.Sum(nil)
-            hasher.Reset()
-        } else {
-            hasher.Write(append(currentHash, node...))
-            currentHash = hasher.Sum(nil)
-            hasher.Reset()
-        }
-    }
-
-    return bytes.Compare(currentHash, root) == 0
+func VerifyProof(proof SparseMerkleProof, root []byte, key []byte, value []byte, hasher hash.Hash) bool {
```

Comment:
Is there a reason `root`, `key`, and `value` aren't part of the `SparseMerkleProof` struct?

Reply:
A Merkle proof is a proof for the statement "a key x has value y for the tree with root z". It doesn't make sense for the statement being proven to be a part of the proof, because the prover could always generate a valid proof if they get to pick the statement. Furthermore, the verifier may already know what the statement being proven is, so including it as part of the proof may be redundant. At minimum, they should already know what the root of the tree is.